### PR TITLE
Add restrict keyword in sincos wrappers.

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
Some oversight.

## What type(s) of changes does this code introduce?
- Bugfix

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

**Comment by ye-luo** [comment]
Start testing in-house

**Comment by ye-luo** [comment]
Start testing in-house

**Comment by PDoakORNL** [comment]
I feel like `restrict` in c++ is often a cargo cult sort of change.  I.e. made in hope that sprinkling it everywhere will result in optimization.  Does this actually cause vectorization passes to work on some compilers?  Do we know it doesn't break any?

File: src/Platforms/CPU/math.hpp [hunk 1]
```diff
@@ -29,36 +29,36 @@ namespace qmcplusplus
...
-inline void sincos(double a, double* s, double* c)
+inline void sincos(double a, double* restrict s, double* restrict c)
 {
   ::sincos(a,s,c);
 }
 
-inline void sincos(float a, float* s, float* c)
+inline void sincos(float a, float* restrict s, float* restrict c)
 {
 #if defined(HAVE_MASS)
```
Comment (PDoakORNL) [inline src/Platforms/CPU/math.hpp:63]: another thing that is strange is there is a vsincos in massv.h

Comment (ye-luo) [reply src/Platforms/CPU/math.hpp:63]: What did you mean?

Comment (PDoakORNL) [reply src/Platforms/CPU/math.hpp:63]: API in mass contains no restrict keyword.

Comment (PDoakORNL) [reply src/Platforms/CPU/math.hpp:63]: I was confused by this comment and the rest of the code being cutoff

**Comment by ye-luo** [comment]
> I feel like `restrict` in c++ is often a cargo cult sort of change. I.e. made in hope that sprinkling it everywhere will result in optimization. Does this actually cause vectorization passes to work on some compilers? Do we know it doesn't break any?

This is a wrapper over C library. sincos always assume no pointer aliasing. So adding restrict is safe.
```
Does this actually cause vectorization passes to work on some compilers?
```
Not exploring this in depth.
```
Do we know it doesn't break any?
```
No breakage observed so far.

File: src/Platforms/CPU/math.hpp (hunk 1 above)
Comment (ye-luo) [reply src/Platforms/CPU/math.hpp:63]: restrict has no impact on the function declaration. It only affects function definition when doing code compilation.

**Comment by PDoakORNL** [comment]
So my only real comment is mass doesn't include restrict in its API. Can't say I really understand it in the context where one is calling an external function with the same pointers but no restrict keywords.

**Comment by ye-luo** [comment]
Even if sincos declaration in math libraries doesn't have restrict keyword, there is nothing wrong for us to suggest restrict. Depending on the compiler, it may inline math functions with fast-math option and thus, Having restrict convey the message of no pointer aliasing.

File: src/Platforms/CPU/math.hpp (hunk 1 above)
Comment (PDoakORNL) [reply src/Platforms/CPU/math.hpp:63]: Alright I think this will cause us no harm. Interested to hear if it helps anywhere.

**Review by PDoakORNL — APPROVED** [review]
LGTM

**Pull request merged**